### ISSUE TITLE
[sklearn#34070] Prevent CPU over-subscription when using thread pools, similar to what process pools do

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#358]
**Mirrored from [scikit-learn#34070](https://github.com/scikit-learn/scikit-learn/issues/34070)**

Original title: Prevent CPU over-subscription when using thread pools, similar to what process pools do
Original labels: New Feature, free-threading, Parallelism and scalability
Created: 2026-05-20T18:37:17Z

---

### Describe the workflow you want to enable

Process pools in sklearn will limit the number of threads used by OpenMP and BLAS. Thread pools will not set any limits.

Consider the following program, which lists how many threads are set for OpenMP/OpenBLAS/etc:

```python
import joblib, sys
from sklearn.utils.parallel import Parallel, delayed
import threadpoolctl

def print_thread_pool_sizes():
    for info in threadpoolctl.threadpool_info():
        print(info["user_api"], info["internal_api"], f"threads: {info['num_threads']}")

with joblib.parallel_config(backend=sys.argv[1]):
    Parallel(n_jobs=-1)([delayed(print_thread_pool_sizes)()])
```

When run under the default backend, the loky process pool, I get:

```
$ python nested-parallelism.py loky
blas openblas threads: 1
blas openblas threads: 1
openmp openmp threads: 1
```

When run under the threading joblib backend, I get:

```
$ python nested-parallelism.py threading
blas openblas threads: 20
blas openblas threads: 20
openmp openmp threads: 20
```

This can lead to significant CPU oversubscription, depending on the semantics of the library:

* When using OpenBLAS, you will get a total of 40 (20 + 20) threads, since the limit is for a single thread pool shared by the whole process.
* When using OpenMP, you will get 400 (20 * 20) threads, since the limit is _per operation starting a thread pool._

The result is code that runs slower, wasting CPU cycles, and not actually achieving any more work.

### Describe your proposed solution

There are a number of scenarios:

**Scenario 1: All cores:** If the number of threads in a thread pool is the same as the number of (available) CPU cores, setting the OpenMP/BLAS limit to 1 will work correctly: no oversubscription, no cores sitting idle.
 
**Scenario 2: All cores but 1 or 2:** If the requested number of threads/processes is negative number `-N` in joblib, this is defined as "number of available cores + 1- N". I.e. this semantically suggests the users wants to saturate the cores, but leave a few cores for other tasks. Typically N will be small, e.g. 2 to leave one core free but use all others. In this case again setting the OpenMP/BLAS limit to 1 is the right thing to do.

**Scenario 3: A hard-coded small number of cores, less than available cores:** The remaining case is most likely when the number of cores is limited due to the way the algorithm works. For process pools joblib will already automatically set the number of threads in OpenMP/BLAS to `max(cpu_count() // effective_n_jobs, 1)`. So for example if you have 10 cores and only starting a pool of size 5, OpenMP/BLAS will each be set to allow 2 threads. In a process pool, this works fine.

For a top-level thread pool, this heuristic works fine for OpenMP but _not_ OpenBLAS, since the meaning of "2" is different: for OpenMP it means "2 threads for every operation started in the 5 threads, for a total of 2 * 5 == 10 threads", for OpenBLAS it means "2 threads for the whole process, for a total 

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [58](https://github.com/godofecht/flow-scikit/issues/58) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [59](https://github.com/godofecht/flow-scikit/issues/59) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
